fecgov/openFEC
# [MERGE WITH GIT FLOW] Hotfix: Feature to separate Schedule A traffic

## Summary (required)

- Resolves #4268

If `SQLA_ROUTE_SCHEDULE_A` is set and there is more than one replica, route non-schedule A traffic to replica 1 and schedule A to everything else.

Original research here: https://github.com/fecgov/openFEC/issues/4038

## How to test the changes locally

- Locally, set `SQLA_FOLLOWERS` to the values from `prod` (has 3 replicas)
- Locally, set `SQLA_ROUTE_SCHEDULE_A` to `True`
- Check out this branch and run the API
- Test out schedule A and non-schedule A queries
- You can add print statements to check to these lines, instead of comments:
```
            if self.route_schedule_a:
                if '/schedule_a/' not in request.path:
                    # Route all non-schedule A traffic to replica 1
                    return self.followers[0]
                else:
                    # Split out Schedule A to remaining replicas
                    return random.choice(self.followers[1:])
```
 From a manual deploy:
```
   2020-03-31T12:01:23.71-0400 [APP/PROC/WEB/0] OUT 127.0.0.1 - - [31/Mar/2020:16:01:23 +0000] "GET /v1/calendar-dates/?per_page=500&sort=start_date&min_start_date=2020-3-31&calendar_category_id=21&calendar_category_id=25&calendar_category_id=26&calendar_category_id=27 
   2020-03-31T12:01:23.85-0400 [APP/PROC/WEB/0] OUT Schedule A not in path, routing to replica 1
...
   2020-03-31T12:01:30.77-0400 [APP/PROC/WEB/0] OUT 127.0.0.1 - - [31/Mar/2020:16:01:30 +0000] "GET /v1/schedules/schedule_a/?sort_hide_null=false&sort_nulls_last=true&sort=-contribution_receipt_date&per_page=30&committee_id=C00412304&is_individual=true&two_year_transaction_period=2020 
   2020-03-31T12:01:30.78-0400 [APP/PROC/WEB/0] OUT Schedule A in path, routing to another replica
```

## Impacted areas of the application
List general components of the application that this PR will affect:

-  SQLAlchemy routing sessions
